API for storage credentials and external locations

vikrantpuppala opened the pull request:
**PR Checklist**

- [ ] A description of the changes is added to the description of this PR.
- [ ] If there is a related issue, make sure it is linked to this PR.
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added or modified a feature, documentation in `docs` is updated

**Description of changes**

<!-- Please state what you've changed and how it might affect the users. -->
API spec for storage credentials and external locations

vikrantpuppala committed:
add api

Signed-off-by: Vikrant Puppala <vikrant.puppala@databricks.com>

vikrantpuppala committed:
Merge remote-tracking branch 'unitycatalog/main' into scel/1-api

vikrantpuppala committed:
cleanup

Signed-off-by: Vikrant Puppala <vikrant.puppala@databricks.com>

tdas commented on api/all.yaml:
```diff
@@ -1132,6 +1132,238 @@ paths:
+          description: The storage credential was successfully deleted.
+          content:
+            application/json:
+              schema: {}
```
arent their failure response codes?

vikrantpuppala replied:
we don't specifically define this for CRUD apis of any entities, for CC this was important since it would define how the commit success/failure flow is handled, but otherwise it would be something new that we add.

tdas commented on api/all.yaml:
```diff
@@ -1132,6 +1132,238 @@ paths:
+      parameters:
+        - name: force
+          in: query
+          description: Force deletion even if there are dependent external locations or external tables.
```
specify explicitly what happens if this optional parameter is not specified and there are dependent stuff

vikrantpuppala replied:
the deletion would fail, do we need to explicitly specify this?

tdas commented on api/all.yaml:
```diff
@@ -1132,6 +1132,238 @@ paths:
+      description: |
+        Gets an array of storage credentials (as __StorageCredentialInfo__ objects).
+        The array is limited to only those storage credentials the caller has permission to access.
+        If the caller is a metastore admin, retrieval of credentials is unrestricted.
```
is the metastore admin role defined in UC OSS?

vikrantpuppala replied:
yes, defined like: https://github.com/vikrantpuppala/unitycatalog/pull/13/files#diff-966f83bb7b88e2a4bdbb5bed92d7be050d6722401a6901bfb065dda368f4a58bR44

tdas commented on api/all.yaml:
```diff
@@ -1132,6 +1132,238 @@ paths:
+            type: string
+          required: false
+      responses:
+        200:
```
are there error codes?

tdas commented on api/all.yaml:
```diff
@@ -2303,6 +2535,214 @@ components:
+      type: object
+      properties:
+        name:
+          description: The credential name. The name must be unique within the metastore.
```
what are the restrictions of a name? are their any? spaces? special chars? 

if spaces are allowed, will get by name work?

vikrantpuppala replied:
here is our validation for names across entities: https://github.com/unitycatalog/unitycatalog/blob/f88028a3f352645016d87830ca118511c0612187/server/src/main/java/io/unitycatalog/server/utils/ValidationUtils.java#L12

tdas commented on api/all.yaml:
```diff
@@ -2303,6 +2535,214 @@ components:
+        name:
+          description: The credential name. The name must be unique within the metastore.
+          type: string
+        aws_iam_role:
```
cant you give descriptions here regarding what this is?
same qs for azure below

vikrantpuppala replied:
properties with` $ref `cannot have `description`, however, we could add another parameter called allOf to provide both as documented here: https://swagger.io/docs/specification/v3_0/data-models/oneof-anyof-allof-not/#allof. This would be something new for UC OSS spec, let me know if you want to go that way

tdas commented on api/all.yaml:
```diff
@@ -2303,6 +2535,214 @@ components:
+          type: string
+        aws_iam_role:
+          $ref: '#/components/schemas/AwsIamRoleResponse'
+        azure_service_principal:
```
lets add azure later, when adding the azure support. so lets start with aws only.

vikrantpuppala replied:
can we add azure support together?

tdas commented on api/all.yaml:
```diff
@@ -2303,6 +2535,214 @@ components:
+          description: Username of current owner of credential.
+          type: string
+        full_name:
+          description: The full name of the credential.
```
what is the difference between name an full_name?

vikrantpuppala replied:
none as of today, they contain the same thing

tdas commented on api/all.yaml:
```diff
@@ -2303,6 +2535,214 @@ components:
+          type: integer
+          format: int64
+        created_by:
+          description: Username of credential creator.
```
does this apply in UC OSS? do users have names?

vikrantpuppala replied:
Username in UC OSS is the email itself. All entities say the same: [TableInfo, for example](https://github.com/unitycatalog/unitycatalog/blob/f88028a3f352645016d87830ca118511c0612187/api/all.yaml#L1354-L1373)

tdas commented on api/all.yaml:
```diff
@@ -2303,6 +2535,214 @@ components:
+          type: integer
+          format: int64
+        updated_by:
+          description: Username of user who last modified the credential.
```
same question as above.

tdas commented on api/all.yaml:
```diff
@@ -2303,6 +2535,214 @@ components:
+          type: string
+        aws_iam_role:
+          $ref: '#/components/schemas/AwsIamRoleRequest'
+        azure_service_principal:
```
same .. dont add azure as the first pass.

tdas commented on api/all.yaml:
```diff
@@ -2303,6 +2535,214 @@ components:
+        unity_catalog_iam_arn:
+          description: The Amazon Resource Name (ARN) of the AWS IAM used by the Unity Catalog Server. This is the identity that is going to assume the AWS IAM role.
+          type: string
+        external_id:
```
what is this????

vikrantpuppala replied:
to solve this: https://docs.aws.amazon.com/IAM/latest/UserGuide/confused-deputy.html

tdas replied:
add this link to the description of the external_id

tdas commented on api/all.yaml:
```diff
@@ -2303,6 +2535,214 @@ components:
+        external_id:
+          description: The external ID used in role assumption to prevent confused deputy problem.
+          type: string
+    AzureServicePrincipal:
```
remove azure

vikrantpuppala replied:
can we push in azure together?

tdas commented on api/all.yaml:
```diff
@@ -2303,6 +2535,214 @@ components:
+        comment:
+          description: User-provided free-form text description.
+          type: string
+        access_point:
```
example values of this?

tdas replied:
why is this needed again?

vikrantpuppala replied:
access points allow creating multiple entry points with unique names, policies, and network controls for a single bucket essentially help simplify bucket policies

vikrantpuppala replied:
however, not needed for initial version

tdas commented on api/all.yaml:
```diff
@@ -2303,6 +2535,214 @@ components:
+        owner:
+          description: The owner of the external location.
+          type: string
+        access_point:
```
example values

vikrantpuppala replied:
not needed for initial version

tdas commented on api/all.yaml:
```diff
@@ -1132,6 +1132,238 @@ paths:
+          schema:
+            type: string
+          required: false
+        - name: include_browse
```
what is this for and can we not do this?

vikrantpuppala replied:
yes, we can skip this in the initial version

tdas commented on api/all.yaml:
```diff
@@ -2303,6 +2535,214 @@ components:
+      type: object
+      properties:
+        name:
+          description: Name of the external location.
```
same questions as name in the Storage Credential

vikrantpuppala committed:
comments

Signed-off-by: Vikrant Puppala <vikrant.puppala@databricks.com>

vikrantpuppala committed:
Merge remote-tracking branch 'refs/remotes/unitycatalog/main' into scel/1-api

vikrantpuppala committed:
update apis to use generalised credentials

Signed-off-by: Vikrant Puppala <vikrant.puppala@databricks.com>

vikrantpuppala committed:
Merge remote-tracking branch 'refs/remotes/unitycatalog/main' into scel/1-api

BenWilson2 committed:
Add functionality for retrieving callable definitions and converting them to callables (#912)

**PR Checklist**

- [X] A description of the changes is added to the description of this
PR.
- [ ] If there is a related issue, make sure it is linked to this PR.
- [ ] If you've fixed a bug or added code that should be tested, add
tests!
- [ ] If you've added or modified a feature, documentation in `docs` is
updated

**Description of changes**

Extension of #911 , bringing a unified functionality across both
clients. Adds a utility for converting the string representation to an
actual callable for local execution.

---------

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

avriiil committed:
[DOCS] Update links to AI integration example notebooks (#893)

This PR updates the broken links to AI integration example notebooks and
adds the links for the ones that were missing.
Will close #892. 

Link to LlamaIndex notebook should be verified. I could not find an OSS
example in the repo.

Signed-off-by: Avril Aysha <68642378+avriiil@users.noreply.github.com>

---------

Signed-off-by: Avril Aysha <68642378+avriiil@users.noreply.github.com>
Signed-off-by: Ben Wilson <39283302+BenWilson2@users.noreply.github.com>
Co-authored-by: Ben Wilson <39283302+BenWilson2@users.noreply.github.com>

tnightengale committed:
[BUILD] Push official docker image on tagged release (#910)

**PR Checklist**

- [x] A description of the changes is added to the description of this
PR.
- [x] If there is a related issue, make sure it is linked to this PR.
- [x] If you've fixed a bug or added code that should be tested, add
tests!
- [x] If you've added or modified a feature, documentation in `docs` is
updated

**Description of changes**

Duplicate of #698

Pushed from non-fork to allow access to Docker secrets, uploaded to this
repository.

<!-- Please state what you've changed and how it might affect the users.
-->

Signed-off-by: Teghan Nightengale <tnightengale@gmail.com>
Co-authored-by: Denny Lee <denny.g.lee@gmail.com>

BenWilson2 committed:
unpin databricks-connect (#928)

**PR Checklist**

- [X] A description of the changes is added to the description of this
PR.
- [ ] If there is a related issue, make sure it is linked to this PR.
- [ ] If you've fixed a bug or added code that should be tested, add
tests!
- [ ] If you've added or modified a feature, documentation in `docs` is
updated

**Description of changes**

Unpins databricks-connect to use a minimum version specifier. Manual
testing in DBR has confirmed that later versions work just fine with
higher versions.

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Increment the crewai integration version for a patch release (#923)

**PR Checklist**

- [ ] A description of the changes is added to the description of this
PR.
- [ ] If there is a related issue, make sure it is linked to this PR.
- [ ] If you've fixed a bug or added code that should be tested, add
tests!
- [ ] If you've added or modified a feature, documentation in `docs` is
updated

**Description of changes**

<!-- Please state what you've changed and how it might affect the users.
-->

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Fix serverless session resumption (#918)

**PR Checklist**

- [X] A description of the changes is added to the description of this
PR.
- [ ] If there is a related issue, make sure it is linked to this PR.
- [ ] If you've fixed a bug or added code that should be tested, add
tests!
- [ ] If you've added or modified a feature, documentation in `docs` is
updated

**Description of changes**

This PR forces the SparkSession (triggered by the call to
`set_default_spark_session` and the call to the builder to reconnect to
the serverless instance in the case of session or cluster invalidation)
to rebuild the spark session in accordance with recommendations of the
serverless team. Also, in the event that the retry logic is exhausted,
provide guidance for notebook users on how they can quickly get back up
and running by detaching and reattaching their notebook to an active
serverless cluster instance.

---------

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Fix Anthropic and OpenAI utils tests (#929)

**PR Checklist**

- [X] A description of the changes is added to the description of this
PR.
- [ ] If there is a related issue, make sure it is linked to this PR.
- [ ] If you've fixed a bug or added code that should be tested, add
tests!
- [ ] If you've added or modified a feature, documentation in `docs` is
updated

**Description of changes**

After merging both the dbconnect pin restriction AND the serverless
retry logic, it turns out that we need another patch for the spark
session initialization in the utils patch fixtures.

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
silence langchain warning (#927)

**PR Checklist**

- [X] A description of the changes is added to the description of this
PR.
- [ ] If there is a related issue, make sure it is linked to this PR.
- [ ] If you've fixed a bug or added code that should be tested, add
tests!
- [ ] If you've added or modified a feature, documentation in `docs` is
updated

**Description of changes**

Silences a known noisy warning when using the LangChain toolkit
integration.

---------

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

vikrantpuppala committed:
Merge remote-tracking branch 'refs/remotes/unitycatalog/main' into scel/1-api

vikrantpuppala committed:
update comment

Signed-off-by: Vikrant Puppala <vikrant.puppala@databricks.com>

BenWilson2 committed:
Local exec (#925)

**PR Checklist**

- [X] A description of the changes is added to the description of this
PR.
- [ ] If there is a related issue, make sure it is linked to this PR.
- [ ] If you've fixed a bug or added code that should be tested, add
tests!
- [ ] If you've added or modified a feature, documentation in `docs` is
updated

**Description of changes**

Adds a local sandbox mode for execution that can protect the running
REPL context (and underlying VM) from computationally complex function
executions and limits the available libraries that can be utilized
within the function execution to provide a modicum of security as
compared to raw execution of a callable in the main process.

---------

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>
Signed-off-by: Ben Wilson <39283302+BenWilson2@users.noreply.github.com>

ccmao1130 committed:
Update Daft Documentation Link (#884)

aravind-segu committed:
Propagate Workspace Client Config to DB Connect (#930)

**PR Checklist**

- [x] A description of the changes is added to the description of this
PR.
- [ ] If there is a related issue, make sure it is linked to this PR.
- [ ] If you've fixed a bug or added code that should be tested, add
tests!
- [ ] If you've added or modified a feature, documentation in `docs` is
updated

**Description of changes**
In this PR, if a client is defined, the config is propagated to the
Databricks Session. This is done by setting the `sdkConfig` parameter.
However this parameter cannot be set when serverless is set to True. So
we first alter the config and set serverless to auto, which has the same
effect as setting `serverless` to True on Databricks Session.

<!-- Please state what you've changed and how it might affect the users.
-->

---------

Signed-off-by: aravind-segu <aravind.segu@databricks.com>

aravind-segu committed:
Uc function client invoker access (#933)

**PR Checklist**

- [ ] A description of the changes is added to the description of this
PR.
- [ ] If there is a related issue, make sure it is linked to this PR.
- [ ] If you've fixed a bug or added code that should be tested, add
tests!
- [ ] If you've added or modified a feature, documentation in `docs` is
updated

This pr introduces a new flag `filter_accessible_functions` to
UCFunctionToolkit. When set to true and a client's get function throws a
permission error, this function is skipped when adding to
UCFunctionToolkit

---------

Signed-off-by: aravind-segu <aravind.segu@databricks.com>

ognis1205 committed:
[UI] Update App to provide AuthProvider only when auth is enabled (#909)

**PR Checklist**

- [x] A description of the changes is added to the description of this
PR.
- [x] If there is a related issue, make sure it is linked to this PR.
- [x] If you've fixed a bug or added code that should be tested, add
tests!
- [x] If you've added or modified a feature, documentation in `docs` is
updated

**Description of changes**

- Updated `App` to provide `AuthProvider` only when authentication is
enabled.

Will close #766 .

Signed-off-by: Shingo OKAWA <shingo.okawa.g.h.c@gmail.com>

ognis1205 committed:
[UI] Add rowKey property to ListLayout (#908)

**PR Checklist**

- [x] A description of the changes is added to the description of this
PR.
- [x] If there is a related issue, make sure it is linked to this PR.
- [x] If you've fixed a bug or added code that should be tested, add
tests!
- [x] If you've added or modified a feature, documentation in `docs` is
updated

**Description of changes**

- Added the `rowKey` property to `ListLayout` to ensure proper row
identification and prevent error messages.
 
 Will close #846 .

Signed-off-by: Shingo OKAWA <shingo.okawa.g.h.c@gmail.com>

ognis1205 committed:
Update auth-related openapi specification. (#771)

**PR Checklist**

- [x] A description of the changes is added to the description of this
PR.
- [x] If there is a related issue, make sure it is linked to this PR.
- [x] If you've fixed a bug or added code that should be tested, add
tests!
- [x] If you've added or modified a feature, documentation in `docs` is
updated

**Description of changes**

- Added models and APIs related to authentication to OpenAPI
specification.
- Updated the server-side logic to align with the updated OpenAPI
specifications.
- Modified the CLI to support the changes introduced in the updated
OpenAPI specification.
 - ~Fixed the server bootstrap order to ensure proper initialization.~

Will close #768 .

---------

Signed-off-by: Shingo OKAWA <shingo.okawa.g.h.c@gmail.com>

avriiil committed:
[DOCS] add client definition to toolkit creation (#915)

When testing the LangChain integration code I ran into the following
error when creating the Toolkit instance following the syntax in the
docs:

`UC function client is not set.`

Adding the `client=client` to define the UC function client solves the
issue. Feedback welcome.

Signed-off-by: Avril Aysha <68642378+avriiil@users.noreply.github.com>
Co-authored-by: Ben Wilson <39283302+BenWilson2@users.noreply.github.com>

BenWilson2 committed:
Remove ancient package dependency (#937)

**PR Checklist**

- [X] A description of the changes is added to the description of this
PR.
- [ ] If there is a related issue, make sure it is linked to this PR.
- [ ] If you've fixed a bug or added code that should be tested, add
tests!
- [ ] If you've added or modified a feature, documentation in `docs` is
updated

**Description of changes**

We shouldn't be including the old asyncio dependency in the build as it
is incompatible with Python 3.13. The functionality of this library was
merged many years ago into the core asyncio built-in provided asyncio
module within base Python.

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

tdas committed:
Merge branch 'main' into scel/api

BenWilson2 committed:
Add default client inference with Toolkit creation (#938)

**PR Checklist**

- [x] A description of the changes is added to the description of this
PR.
- [ ] If there is a related issue, make sure it is linked to this PR.
- [ ] If you've fixed a bug or added code that should be tested, add
tests!
- [ ] If you've added or modified a feature, documentation in `docs` is
updated

**Description of changes**

Adds the ability to assume a default serverless client for
UCFunctionToolkit instances if the required connectivity packages are
installed in the system.

---------

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

serena-ruan committed:
Set default client in get_uc_function_client (#940)

**PR Checklist**

- [ ] A description of the changes is added to the description of this
PR.
- [ ] If there is a related issue, make sure it is linked to this PR.
- [ ] If you've fixed a bug or added code that should be tested, add
tests!
- [ ] If you've added or modified a feature, documentation in `docs` is
updated

**Description of changes**
We don't need a warning for get_uc_function_client when it's none. It's
also confusing when it's used in validate_or_set_default_client.
This PR also moves the logic to set default client in
get_uc_function_client, so users don't need to create a
UCFunctionToolkit in order to trigger default client setup.

<!-- Please state what you've changed and how it might affect the users.
-->

---------

Signed-off-by: serena-ruan <serena.rxy@gmail.com>

BenWilson2 committed:
Release notes and version update for unitycatalog-ai 0.3.0 (#939)

**PR Checklist**

- [ ] A description of the changes is added to the description of this
PR.
- [ ] If there is a related issue, make sure it is linked to this PR.
- [ ] If you've fixed a bug or added code that should be tested, add
tests!
- [ ] If you've added or modified a feature, documentation in `docs` is
updated

**Description of changes**

<!-- Please state what you've changed and how it might affect the users.
-->

---------

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

prrao87 committed:
Add Kuzu database integration to Unity Catalog docs (#926)

This PR adds [Kuzu](https://kuzudb.com/), an embedded graph database, to
the integration docs. Kuzu's latest release supports scanning from Delta
tables that are stored in Unity Catalog. These docs help Kuzu users
integrate their workflows with Unity Catalog.

**PR Checklist**

- [x] A description of the changes is added to the description of this
PR.
- [x] If there is a related issue, make sure it is linked to this PR.
- [x] If you've fixed a bug or added code that should be tested, add
tests!
- [x] If you've added or modified a feature, documentation in `docs` is
updated

**Description of changes**

Single new markdown file `unity-catalog-kuzu.md` in the
`/docs/integrations` location where other providers' docs are also
placed.

---------

Signed-off-by: Prashanth Rao <35005448+prrao87@users.noreply.github.com>

tnightengale committed:
fix: only login to docker hub on tag push (#936)

**PR Checklist**

- [x] A description of the changes is added to the description of this
PR.
- [x] If there is a related issue, make sure it is linked to this PR.
- [x] If you've fixed a bug or added code that should be tested, add
tests!
- [x] If you've added or modified a feature, documentation in `docs` is
updated

**Description of changes**

Fixes the `docker-build.yml` GitHub workflow to only login to DockerHub
on `push` with tags, ie. a release by a maintainer. Previously was
failing because repository secrets are not available in PRs from forks.
Also uses Github Actions cache for image builds, because DockerHub would
require login.

---------

Signed-off-by: Teghan Nightengale <tnightengale@gmail.com>

vikrantpuppala committed:
Merge remote-tracking branch 'refs/remotes/unitycatalog/main' into scel/1-api

vikrantpuppala committed:
Merge remote-tracking branch 'origin/scel/api' into scel/1-api
